### PR TITLE
Migrate element instances marked as parallel multi instance

### LINKED ISSUE #22129: Migrate parallel multi-instance element instances
## Description

Following cases should be supported:

### Migrate multi-instance service tasks

Screenshot 2024-09-10 at 18 19 45

A multi-instance service task itself, its instances and created jobs (per instance) should be migrated. A mapping should be provided between A->B.

### Migrate multi-instance receive tasks

Screenshot 2024-09-10 at 18 20 33

A multi-instance receive task itself, its instances and created message subscriptions (per instance) should be migrated. A mapping should be provided between A->B.

### Migrate multi-instance embedded subprocess

Screenshot 2024-09-10 at 18 23 30

A multi-instance embedded subprocess itself, its instances and any active element inside instances should be migrated. A mapping should be provided between sub1->sub2 and A->B.

### Migrate multi-instance call activities

Screenshot 2024-09-10 at 18 24 21

A multi-instance call activity should be migrated. A mapping should be provided between A->B. 

Please note that it is not possible to migrate each call activity instance because they belong to another process instance. A new migration should be defined per called activity to migrate each child instance because they are separate process instances.

####  Notes
- Migrating multi-instance body with some of its instances are completed should be possible as well. For example, if 2 instances out of 3 is completed for a multi-instance service task, remaining active instance should be migrated.
- Migrating large set of multi instances (exceeding batch size) is left out of scope. Migrating such multi instances will exceed `MAX_MESSAGE_SIZE` and fail.

### PR DESCRIPTION
**Pull request by berkaycanbc**
## Description

Following cases should be supported:

### Migrate multi-instance service tasks

Screenshot 2024-09-10 at 18 19 45

A multi-instance service task itself, its instances and created jobs (per instance) should be migrated. A mapping should be provided between A->B.

### Migrate multi-instance receive tasks

Screenshot 2024-09-10 at 18 20 33

A multi-instance receive task itself, its instances and created message subscriptions (per instance) should be migrated. A mapping should be provided between A->B.

### Migrate multi-instance embedded subprocess

Screenshot 2024-09-10 at 18 23 30

A multi-instance embedded subprocess itself, its instances and any active element inside instances should be migrated. A mapping should be provided between sub1->sub2 and A->B.

### Migrate multi-instance call activities

Screenshot 2024-09-10 at 18 24 21

A multi-instance call activity should be migrated. A mapping should be provided between A->B. 

Please note that it is not possible to migrate each call activity instance because they belong to another process instance. A new migration should be defined per called activity to migrate each child instance because they are separate process instances.

####  Notes
- Migrating multi-instance body with some of its instances are completed should be possible as well. For example, if 2 instances out of 3 is completed for a multi-instance service task, remaining active instance should be migrated.
- Migrating large set of multi instances (exceeding batch size) is left out of scope. Migrating such multi instances will exceed `MAX_MESSAGE_SIZE` and fail.

## Related issues

closes #22129

### CODE REVIEW

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateMultiInstanceBodyTest.java [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+                Bpmn.createExecutableProcess(processId)
+                    .startEvent()
+                    .serviceTask(
+                        "serviceTask1",
+                        t ->
+                            t.zeebeJobTypeExpression("jobType")
+                                .multiInstance(
+                                    b ->
+                                        b.zeebeInputCollectionExpression("jobTypes")
+                                            .zeebeInputElement("jobType")))
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateMultiInstanceBodyTest.java:59]: 💭 Interesting construct to create jobs with types based on the input collection. This is not the common usage though. Most often, users simply need to a single task multiple times. Cool that this works though! And I can see that it makes the test easier to verify

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateMultiInstanceBodyTest.java [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+    engine.job().ofInstance(childProcessInstanceKey).withType("A").complete();
+
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+                .withProcessInstanceKey(childProcessInstanceKey)
+                .withElementType(BpmnElementType.END_EVENT)
+                .withElementId("multi_instance_child_process_end")
+                .findAny())
+        .describedAs("Expect that the process instance is continued in the target process")
+        .isPresent();
```
Comment (korthout) [inline zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateMultiInstanceBodyTest.java:566]: ❌ This verifies that the end event of the child process instance has activated. However, the child process instance was not migrated. We need to verify that the parent process instance continues in the target process, e.g.:
- after all jobs are completed, the multi-instance body is completed and flows to a different element (one that didn't exist in the source).

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java [hunk 3]
```diff
@@ -451,20 +452,37 @@ public static void requireSameElementType(
...
+    if (targetElementType == BpmnElementType.MULTI_INSTANCE_BODY) {
+      final ExecutableMultiInstanceBody targetElement =
+          targetProcessDefinition
+              .getProcess()
+              .getElementById(targetElementId, ExecutableMultiInstanceBody.class);
+      targetElementType = targetElement.getInnerActivity().getElementType();
+      if (elementInstanceRecord.getBpmnElementType() == targetElementType) {
+        return;
+      }
+    }
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:474]: ❌ Please add a test case that shows this case. E.g. a multi-instance marked service task changed to a multi-instance marked user task.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java [hunk 4]
```diff
@@ -540,24 +558,38 @@ public static void requireUnchangedFlowScope(
...
+        final ExecutableMultiInstanceBody targetElement =
+            targetProcessDefinition
+                .getProcess()
+                .getElementById(targetElementId, ExecutableMultiInstanceBody.class);
+
+        actualFlowScopeId = targetElement.getInnerActivity().getFlowScope().getId();
+        if (expectedFlowScopeId.equals(actualFlowScopeId)) {
+          return;
+        }
+      }
```
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceMigrationPreconditions.java:582]: ❌ Let's add a test case that cover this.

**Review by korthout — CHANGES_REQUESTED** [review]
Great work @berkaycanbc 🚧 

❌ There are some edge cases that are not covered yet. Please add tests for them to show that we can handle them correctly.

For example, we need to add a test that showcases that you can't change the element type of the multi-instance.

We also need to cover rejections for the following edge cases:
- a sequential multi-instance marked activity cannot be changed to parallel multi-instance
- a parallel multi-instance marked activity cannot be changed to sequential multi-instance
- a multi-instance marked activity cannot be unmarked as multi-instance
- a non-multi-instance activity cannot be marked as multi-instance

Lastly, we need to show that incidents caused by the `output element`, `output collection`, and `completion condition` expressions can be resolved through instance migration.

💭 We've already discussed that sequential multi-instance will be covered separately.

**Review by korthout — COMMENTED** [review]
Great stuff @berkaycanbc 

🙇 I was unable to complete the comprehensive review

I have left a few comments for you to consider during my short absence. I'll complete my comprehensive review Tuesday. Please have a look at my comments in the meantime, if you find the time.

💭 I haven't covered all tests in my review
💭 I haven't checked whether all edge cases are now covered by tests

👍 However, generally this looks good

🔧 You could consider testing migrations that map elements nested inside a multi-instance embedded sub-process changing the flow scope. While this should probably already be covered by our checks, I have a feeling that our checks may not cover them well in the same way that you found in 4f11ab6c94dbdd4827f717d6b6e27f2ca218b528:
- a user task inside a multi-instance embedded sub-process is mapped to a user task inside a doubly nested embedded sub-process (i.e. a sub-process is added between the inner instance and the user task), we should reject this case as the flow scope has changed
- a user task inside a multi-instance embedded sub-process is mapped to a user task outside of the embedded sub-process (i.e. it is moved out of the subprocess), we should reject this case as the flow scope has changed

**Comment by korthout** [comment]
Rebased on `main` (resolving a small conflict in 08d976e6ac80040bcb15fe26eb881ded2e6ff91b) and added the two requested test cases.

**Comment by korthout** [comment]
Pair review with @berkaycanbc:
- one more test needed for changing output collection during migration
- one concurrent test: activate parallel multi-instance body, followed by migration, before `ProcessInstanceBatch:ACTIVATE_ELEMENT`

**Comment by korthout** [comment]
@berkaycanbc I had to add a fix in 3c0fa11 due to the output collection changed case. For the rest all is as expected and discussed.

**Review by korthout — APPROVED** [review]
